app/src/actioncommands.cpp

```diff
@@ -563,6 +568,186 @@ void ActionCommands::duplicateKey()
     mEditor->layers()->notifyAnimationLengthChanged();
 }
 
+
+void ActionCommands::manipulateRange()
+{
+    int loopBegin = 1, loopEnd = 2; // default values
+    if (mEditor->playback()->isRangedPlaybackOn())
+    {   // defined range values if available
+        loopBegin = mEditor->playback()->getRangedStartFrame();
+        loopEnd = mEditor->playback()->getRangedEndFrame();
+    }
+    LayerManager* layerMgr = mEditor->layers();
+    int lIndex = layerMgr->currentLayer()->type();
+    //   BITMAP = 1     VECTOR = 2  //not yet      SOUND = 4     CAMERA = 5
+    if (lIndex == 1 || lIndex == 2) //not yet || lIndex == 4 || lIndex == 5)
+    {
+        CopyMultiplekeyframesDialog* cd = new CopyMultiplekeyframesDialog(layerMgr, loopBegin, loopEnd, new QWidget);
+        cd->exec();
+
+        // If validation is NOT OK -> return
+        if (cd->result() == QDialog::Accepted && !cd->getValidity())
+        {
+            QMessageBox msgBox;
+            msgBox.setWindowTitle(tr("Action not valid!"));
+            msgBox.setText(tr("Failed validation check:\n- Exceeds 9999 Frames OR\n- Range not valid"));
+            msgBox.exec();
+            return;
+        }
+
+        // If Validation is OK
+        QString sel = cd->getActiveTab();
+        if (cd->result() == QDialog::Accepted)
+        {
+            int startL = cd->getFirstFrame();
+            int stopL = cd->getLastFrame();
+            QString strFromLayer = cd->getFromLayer();
+            Layer *fromLayer = layerMgr->findLayerByName(strFromLayer);
```

Comment:
Q_CHECK_PTR could be used instead. I'm not sure which one should be preferred.

Reply:
Didn’t even know that one existed. But from it’s documentation:

> If pointer is 0, prints a message containing the source code's file name and line number, saying that the program ran out of memory and aborts program execution. It throws std::bad_alloc instead if exceptions are enabled.

To me this sounds like an assertion would be more appropriate here, since we’re not allocating any memory and we don’t want an exception. That said, gotta keep this in mind for the future.